lib/browser/api/menu.js

```diff
@@ -272,71 +127,154 @@ Menu.prototype.insert = function (pos, item) {
-    if (item.submenu != null) {
-      item.submenu._callMenuWillShow()
-    }
+  if (this.delegate != null) this.delegate.menuWillShow()
+  this.items.forEach(item => {
+    if (item.submenu != null) item.submenu._callMenuWillShow()
   })
 }
 
-var applicationMenu = null
+/* Static Methods */
 
+// return application menu
+Menu.getApplicationMenu = () => applicationMenu
+
+Menu.sendActionToFirstResponder = bindings.sendActionToFirstResponder
+
+// set application menu with a preexisting menu
 Menu.setApplicationMenu = function (menu) {
   if (!(menu === null || menu.constructor === Menu)) {
     throw new TypeError('Invalid menu')
   }
 
-  // Keep a reference.
   applicationMenu = menu
   if (process.platform === 'darwin') {
-    if (menu === null) {
-      return
-    }
+    if (menu === null) return
     menu._callMenuWillShow()
     bindings.setApplicationMenu(menu)
   } else {
-    BrowserWindow.getAllWindows().forEach(function (window) {
-      window.setMenu(menu)
-    })
+    const windows = BrowserWindow.getAllWindows()
+    return windows.map(w => w.setMenu(menu))
   }
 }
 
-Menu.getApplicationMenu = function () {
-  return applicationMenu
+// build a menu by passing in a template
+Menu.buildFromTemplate = function (template) {
+  if (!(template instanceof Array)) {
+    throw new TypeError('Invalid template for Menu')
+  }
+
+  const menu = new Menu()
+  const positioned = []
+  let idx = 0
+
+  // sort template by position
+  template.forEach(item => {
+    idx = (item.position) ? indexToInsertByPosition(positioned, item.position) : idx += 1
+    positioned.splice(idx, 0, item)
+  })
+
+  // add each item from positioned menu to application menu
+  positioned.forEach((item) => {
+    if (typeof item !== 'object') {
+      throw new TypeError('Invalid template for MenuItem')
+    }
+    menu.append(new MenuItem(item))
+  })
+
+  return menu
 }
 
-Menu.sendActionToFirstResponder = bindings.sendActionToFirstResponder
+/* Helper Functions */
 
-Menu.buildFromTemplate = function (template) {
-  var insertIndex, item, j, k, len, len1, menu, menuItem, positionedTemplate
-  if (!Array.isArray(template)) {
-    throw new TypeError('Invalid template for Menu')
+// Search between separators to find a radio menu item and return its group id
+function generateGroupId (items, pos) {
+  if (pos > 0) {
+    for (let idx = pos - 1; idx >= 0; idx--) {
+      if (items[idx].type === 'radio') return items[idx].groupId
+      if (items[idx].type === 'separator') break
+    }
+  } else if (pos < items.length) {
+    for (let idx = pos; idx <= items.length - 1; idx++) {
+      if (items[idx].type === 'radio') return items[idx].groupId
+      if (items[idx].type === 'separator') break
+    }
+  }
+  groupIdIndex += 1
+  return groupIdIndex
+}
+
+// Returns the index of item according to |id|.
+function indexOfItemById (items, id) {
+  const foundItem = items.find(item => item.id === id) || null
+  return items.indexOf(foundItem)
```

Comment:
Nit, but if you manage to get a `null` into `items`, it'll manage to return that index even if the`id` can't be found. You might want to just return `-1` if you can't find the item.